packages/keyring-controller/src/KeyringController.ts

```diff
@@ -187,25 +213,37 @@ export class KeyringController extends BaseController<
       setSelectedAddress: PreferencesController['setSelectedAddress'];
       setAccountLabel?: PreferencesController['setAccountLabel'];
     },
-    config?: Partial<KeyringConfig>,
-    state?: Partial<KeyringState>,
+    messenger: KeyringControllerMessenger,
+    config?: KeyringControllerConfig,
+    state?: Partial<KeyringControllerState>,
   ) {
-    super(config, state);
+    super({
+      name,
+      metadata: {
+        vault: { persist: false, anonymous: false },
+        isUnlocked: { persist: false, anonymous: false },
+        keyringTypes: { persist: false, anonymous: false },
+        keyrings: { persist: false, anonymous: false },
+      },
+      messenger,
+      state: {
+        ...defaultState,
+        ...state,
+      },
+    });
+
     this.#keyring = new EthKeyringController(
       Object.assign({ initState: state }, config),
     );
+    this.#keyring.memStore.subscribe(this.#fullUpdate.bind(this));
```

Comment:
@mcmire If the observable callback is synchronous then it will be executed right after the event has been emitted, so an eventual assertion on the state right after (as example) calling `setLocker` would find the state already updated.
Your exact example is covered by this test case: https://github.com/MetaMask/core/pull/1378/files#diff-0d6a6cf9b4604f78e899c932eb8dcbd38b29a5c80b11e235478f9f9c90cdfba2R318

In general, I changed all test cases to assert on the state of the controller, instead of function returns, so we can test exactly what you are concerned of.

Reply:
I just noticed that `setLocked` is sync, so might not be the right example, but we can take this other test case as example: https://github.com/MetaMask/core/pull/1378/files#diff-0d6a6cf9b4604f78e899c932eb8dcbd38b29a5c80b11e235478f9f9c90cdfba2R594-R595

Here `removeAccount` is awaited and the update callback is called before the promise resolution